src/main/java/com/microsoft/sqlserver/jdbc/ConfigurableRetryLogic.java

```diff
@@ -0,0 +1,218 @@
+
+
+/**
+ * Allows configurable statement retry through the use of the 'retryExec' connection property. Each rule read in is
+ * converted to ConfigRetryRule objects, which are stored and referenced during statement retry.
+ *
+ */
+public class ConfigurableRetryLogic {
+    private final static int INTERVAL_BETWEEN_READS_IN_MS = 30000;
+    private final static String DEFAULT_PROPS_FILE = "mssql-jdbc.properties";
+    private static final java.util.logging.Logger CONFIGURABLE_RETRY_LOGGER = java.util.logging.Logger
+            .getLogger("com.microsoft.sqlserver.jdbc.ConfigurableRetryLogic");
+    private static ConfigurableRetryLogic driverInstance = null;
+    private static long timeLastModified;
+    private static long timeLastRead;
+    private static String lastQuery = ""; // The last query executed (used when rule is process-dependent)
+    private static String prevRulesFromConnectionString = "";
+    private static HashMap<Integer, ConfigRetryRule> stmtRules = new HashMap<>();
+    private static final Lock CRL_LOCK = new ReentrantLock();
+    private static final String SEMI_COLON = ";";
+    private static final String COMMA = ",";
+    private static final String FORWARD_SLASH = "/";
+    private static final String EQUALS_SIGN = "=";
+    private static final String RETRY_EXEC = "retryExec";
+
+    private ConfigurableRetryLogic() throws SQLServerException {
+        timeLastRead = new Date().getTime();
+        setUpRules(null);
+    }
+
+    /**
+     * Fetches the static instance of ConfigurableRetryLogic, instantiating it if it hasn't already been.
+     * Each time the instance is fetched, we check if a re-read is needed, and do so if properties should be re-read.
+     *
+     * @return The static instance of ConfigurableRetryLogic
+     * @throws SQLServerException
+     *         an exception
+     */
+    public static ConfigurableRetryLogic getInstance() throws SQLServerException {
+        if (driverInstance == null) {
+            CRL_LOCK.lock();
+            try {
+                if (driverInstance == null) {
+                    driverInstance = new ConfigurableRetryLogic();
+                } else {
+                    refreshRuleSet();
+                }
+            } finally {
+                CRL_LOCK.unlock();
+            }
+        } else {
+            refreshRuleSet();
+        }
+        return driverInstance;
+    }
+
+    /**
+     * If it has been INTERVAL_BETWEEN_READS_IN_MS (30 secs) since last read, see if we last did a file read, if so
+     * only reread if the file has been modified. If no file read, set up rules using the prev. connection string rules.
+     *
+     * @throws SQLServerException
+     *         when an exception occurs
+     */
+    private static void refreshRuleSet() throws SQLServerException {
+        long currentTime = new Date().getTime();
+        if ((currentTime - timeLastRead) >= INTERVAL_BETWEEN_READS_IN_MS) {
+            // If it has been 30 secs, reread

```

Comment:
the value may change, do not mention hardcoded value.
Just something like "if it has been > an interval wait time"